### PR TITLE
fix(nextly): handle empty encoded bodies and over-limit url attachments in safeFetch

### PR DESCRIPTION
**Pull request by mobeenabdullah**
## What

Two follow-up fixes for edge cases in the IP-pinning `safeFetch` change (#232, merged), both flagged by Codex in post-merge review.

### 1. Empty encoded bodies no longer fail
A `2xx`/`204`/`304` response that carries a `Content-Encoding: gzip|br|deflate` header but an **empty body** was turned into a `SafeFetchError` — inflating zero bytes throws. The old `fetch()` path returned an OK empty body, and the form-builder webhook handler only checks `response.ok`, so valid deliveries were being reported as failed. `decodeBody` now passes an empty body straight through.

### 2. Over-limit URL attachments produce the size-validation error
Capping the email attachment fetch at the attachment limit (from an earlier #232 fix) meant an over-limit URL-backed attachment threw `response-too-large` inside `readBytes`, which the resolver wrapped as `EMAIL_ATTACHMENT_STORAGE_READ_FAILED` (a 500-style storage failure) instead of the `EMAIL_ATTACHMENT_SIZE_EXCEEDED` validation error that local/S3 reads produce. Now:
- the attachment fetch translates a `response-too-large` `SafeFetchError` into the size-exceeded validation error, and
- the attachment resolver passes a typed `NextlyError` from `readBytes` through unchanged, wrapping only opaque (non-`NextlyError`) storage failures.

So URL, local, and S3 attachments all fail over-limit the same way.

## Tests

- `safeFetch`: an empty gzip-headed 2xx returns `ok` with an empty body.
- `attachment-resolver`: a typed `NextlyError` from `readBytes` surfaces as-is (not re-wrapped); the existing plain-`Error` case still becomes `STORAGE_READ_FAILED`.
- `pnpm --filter nextly lint` + `check-types` + the email attachment suites pass (36 tests green).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved email attachment error handling by consistently enforcing attachment size limits and returning clearer validation vs internal failure details.
  * Updated attachment resolution to preserve validation errors (oversized attachments) while wrapping unexpected failures as internal errors with improved diagnostics.
  * Fixed external URL content decoding for empty gzip responses and improved “response-too-large” detection for oversized decompressed payloads.

* **Tests**
  * Added regression tests covering attachment size/error propagation and gzip/deflate decompression edge cases, including empty bodies and oversize decompressed content.

### CODE REVIEW

File: packages/nextly/src/di/registrations/register-email.ts [hunk 1]
```diff
@@ -83,16 +85,46 @@ export function registerEmailServices(ctx: RegistrationContext): void {
...
-              `Failed to fetch attachment from ${url}: HTTP ${response.status}`
-            );
+            throw NextlyError.internal({
+              logContext: {
+                emailAttachmentCode:
+                  EmailErrorCode.ATTACHMENT_STORAGE_READ_FAILED,
+                url,
+                status: response.status,
+              },
+            });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/di/registrations/register-email.ts:127]: Good call, fixed in 65fbc993. decodeBody now inspects the error code: node:zlib throws ERR_BUFFER_TOO_LARGE when the inflated output exceeds maxOutputLength (a decompression bomb), which is genuinely a size problem, so it now reports reason "response-too-large" rather than "decode-failed". The email attachment path's existing response-too-large handler then maps it to ATTACHMENT_SIZE_EXCEEDED, so a gzip'd over-limit attachment surfaces the same validation error as a raw one. Genuinely corrupt streams still return decode-failed. Added a test with a body that inflates past the cap asserting reason response-too-large.

File: packages/nextly/src/utils/validate-external-url.ts [hunk 2]
```diff
@@ -621,6 +621,9 @@ function decodeBody(
   url: string
 ): { body: Buffer; decoded: boolean } | SafeFetchError {
   if (!encoding) return { body: buf, decoded: false };
+  // An empty body has nothing to decode; inflating zero bytes would throw and
+  // turn a valid empty 2xx/204/304 response into a failure. Pass it through.
+  if (buf.length === 0) return { body: buf, decoded: false };
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/utils/validate-external-url.ts]: Fixed in d9f047f5. The empty-body shortcut now returns decoded:true, so toWhatwgResponse strips the content-encoding/-length headers just like a normally-decoded body. Updated the empty-body test to assert content-encoding is null.

File: packages/nextly/src/domains/email/services/attachment-resolver.ts [hunk 3]
```diff
@@ -100,6 +100,10 @@ export async function resolveAttachments(
     try {
       content = await deps.readBytes(media.filename);
     } catch (err) {
+      // A typed NextlyError from readBytes (e.g. the size-exceeded validation a
+      // URL-backed fetch raises when the object is over the limit) is already
+      // precise; pass it through and wrap only opaque storage failures.
+      if (NextlyError.is(err)) throw err;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/email/services/attachment-resolver.ts]: Good catch, fixed in d9f047f5. The pass-through is now restricted to NextlyError with code VALIDATION_ERROR (the size-exceeded error), so ExternalUrlError (EXTERNAL_URL_BLOCKED, with its 'Resolved to non-public IP' message) and SafeFetchError (EXTERNAL_REQUEST_FAILED) are wrapped as STORAGE_READ_FAILED again and never leak to the caller. Added a resolver test asserting a non-validation NextlyError is wrapped.

[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/email/services/attachment-resolver.ts]: Fixed in d9f047f5 (same change as the related thread): the resolver now passes through only VALIDATION_ERROR NextlyErrors, so the missing-local-file case (relative URL to safeFetch -> ExternalUrlError) is wrapped as EMAIL_ATTACHMENT_STORAGE_READ_FAILED again, matching the documented behavior. Only the size-exceeded validation error passes through.

File: packages/nextly/src/di/registrations/register-email.ts [hunk 4]
```diff
@@ -83,16 +85,46 @@ export function registerEmailServices(ctx: RegistrationContext): void {
...
+          let response: Response;
+          try {
+            response = await safeFetch(url, {
+              maxResponseBytes: limits.maxTotalBytes,
+            });
+          } catch (err) {
+            if (
+              err instanceof SafeFetchError &&
+              err.reason === "response-too-large"
+            ) {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/di/registrations/register-email.ts:102]: This is addressed by the latest commit on the branch: decodeBody now classifies the zlib maxOutputLength overflow (ERR_BUFFER_TOO_LARGE) as reason 'response-too-large' rather than 'decode-failed', so an encoded attachment whose decoded size exceeds maxTotalBytes hits the response-too-large branch and surfaces as EMAIL_ATTACHMENT_SIZE_EXCEEDED, consistent with local and unencoded URL attachments. Added a test for the inflate-past-cap case.

File: packages/nextly/src/utils/validate-external-url.ts [hunk 5]
```diff
@@ -661,16 +664,31 @@ function decodeBody(
...
     return new SafeFetchError(
-      `Failed to decode ${encoding} response body`,
+      tooLarge
+        ? `Decoded ${encoding} response body exceeded the size cap`
+        : `Failed to decode ${encoding} response body`,
       url,
-      "decode-failed"
+      tooLarge ? "response-too-large" : "decode-failed"
     );
   }
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/utils/validate-external-url.ts:687]: Good catch, fixed in 1df710f9. The deflate raw-fallback now re-throws when the first error is ERR_BUFFER_TOO_LARGE (a size-cap overflow), so it keeps the response-too-large signal; it only retries inflateRawSync for a genuine zlib-wrapped/raw format mismatch. Added a test with a zlib-wrapped deflate body that inflates past the cap asserting reason response-too-large.
